### PR TITLE
Add Attribute structure and decoder

### PR DESCRIPTION
**Pull request by lukewestby**
Formalizing some internal stuff from https://github.com/eeue56/elm-html-test/pull/35 as requested. This also upgrades the elm-test setup and there are a few comment and whitespace changes from elm-format as well.

### CODE REVIEW

File: src/ElmHtml/InternalTypes.elm [hunk 1]
```diff
@@ -8,17 +8,24 @@ module ElmHtml.InternalTypes
...
         , emptyFacts
+        , decodeAttribute
         , ElementKind(..)
         , toElementKind
         )
 
 {-| Internal types used to represent Elm Html in pure Elm
 
-@docs ElmHtml, TextTagRecord, NodeRecord, CustomNodeRecord, MarkdownNodeRecord, Facts, Tagger, EventHandler, ElementKind
+@docs ElmHtml, TextTagRecord, NodeRecord, CustomNodeRecord, MarkdownNodeRecord, Facts, Tagger, EventHandler, ElementKind, Attribute, AttributeRecord, NamespacedAttributeRecord, PropertyRecord, EventRecord
```
Comment (eeue56) [inline src/ElmHtml/InternalTypes.elm]: 🎨 perhaps these can be split up over multiple lines? It's getting a bit hectic

File: src/ElmHtml/InternalTypes.elm [hunk 2]
```diff
@@ -125,10 +139,61 @@ type HtmlContext msg
     = HtmlContext (List Tagger) (List Tagger -> EventHandler -> Json.Decode.Decoder msg)
 
 
+{-| Type for representing Elm's Attributes
+
+  - Attribute is an HTML attribute, like `Html.Attributes.title`
```
Comment (eeue56) [inline src/ElmHtml/InternalTypes.elm]: It's a bit hard to tell from this explantation what attribute vs property is, but NBD

File: src/ElmHtml/InternalTypes.elm [hunk 3]
```diff
@@ -125,10 +139,61 @@ type HtmlContext msg
...
+    , decoder : Json.Decode.Decoder msg
+    , options : Html.Events.Options
+    }
+
+
 {-| decode a json object into ElmHtml, you have to pass a function that decodes
 events from Html Nodes. If you don't want to decode event msgs, you can ignore it:
 
-decodeElmHtml (\_ _ -> ()) jsonHtml
+decodeElmHtml (_ _ -> ()) jsonHtml
```
Comment (eeue56) [inline src/ElmHtml/InternalTypes.elm]: 🤔 pretty sure the backslash shoudl still be there

Comment (lukewestby) [reply src/ElmHtml/InternalTypes.elm]: oh weird, must be a bug in elm-format or elmjutsu or something

File: src/ElmHtml/InternalTypes.elm [hunk 4]
```diff
@@ -387,18 +532,18 @@ voidElements =
...
     [ "script", "style" ]
 
 
 {-| A list of all Escapable Raw Text Elements as defined by the HTML5
-   specification. They can have text and character references, but the text must
-   not contain an ambiguous ampersand along with addional restrictions:
-   https://html.spec.whatwg.org/multipage/syntax.html#cdata-rcdata-restrictions
+specification. They can have text and character references, but the text must
+not contain an ambiguous ampersand along with addional restrictions:
+<https://html.spec.whatwg.org/multipage/syntax.html#cdata-rcdata-restrictions>
```
Comment (eeue56) [inline src/ElmHtml/InternalTypes.elm:554]: Do these angle brackets do anything?

Comment (lukewestby) [reply src/ElmHtml/InternalTypes.elm:554]: elm-format added them

https://github.com/avh4/elm-format/blob/master/src/ElmFormat/Render/Markdown.hs#L179

File: src/ElmHtml/InternalTypes.elm (hunk 1 above)
Comment (lukewestby) [reply src/ElmHtml/InternalTypes.elm]: ✅

File: src/ElmHtml/InternalTypes.elm (hunk 2 above)
Comment (lukewestby) [reply src/ElmHtml/InternalTypes.elm]: ✅

File: src/ElmHtml/InternalTypes.elm (hunk 3 above)
Comment (lukewestby) [reply src/ElmHtml/InternalTypes.elm]: ✅

File: src/ElmHtml/InternalTypes.elm (hunk 4 above)
Comment (eeue56) [reply src/ElmHtml/InternalTypes.elm:554]: Hmm, I'm surprised if that works, but TIL if it does

File: src/ElmHtml/InternalTypes.elm [hunk 5]
```diff
@@ -125,6 +143,61 @@ type HtmlContext msg
...
+-}
+type alias PropertyRecord =
+    { key : String
+    , value : Json.Decode.Value
+    }
+
+
+{-| Event contains a string key, a decoder for a msg and event options
+-}
+type alias EventRecord msg =
```
Comment (eeue56) [inline src/ElmHtml/InternalTypes.elm]: I don't think it's type-safe for us to get `msg` here. The info we get from elm-html-test is just `Json.Value` -- this breaks the type. Best to encode it as Json.Value and make it unparsable

Comment (lukewestby) [reply src/ElmHtml/InternalTypes.elm]: ✅

File: src/ElmHtml/InternalTypes.elm (hunk 3 above)
Comment (lukewestby) [reply src/ElmHtml/InternalTypes.elm]: ✅  for real this time

**Comment by eeue56** [comment]
@lukewestby thanks! published as 5.1.0
